File: src/runtime/config/schemas/decode_policy.cpp
// Registration for the "decode_policy" namespace schema.
//
// Mirrors trtf_build/trtf_build/runtime_config/schemas/decode_policy.py.
// Only edit both sides together; the cross-language match test gates on it.

#include "trtf/config/schemas/decode_policy.h"

#include <any>
#include <set>
#include <string>

namespace trtf::config::schemas {

Schema make_decode_policy_schema()
{
    const std::set<Layer> build_and_bundle = {
        Layer::BuildTime, Layer::BundleDefault,
    };
    return Schema{
        "decode_policy",
        {
            // Build-time only. Chosen when the bundle is built and baked
            // into the engine graph — session/platform layers cannot
            // retroactively toggle it.
            ConfigField{
                "force_manual_attention", "bool",
                std::any{false}, build_and_bundle, nullptr,
            },
        },
    };
}

namespace {
struct DecodePolicyRegistrar {
    DecodePolicyRegistrar()
    {
        SchemaRegistry::instance().register_schema(make_decode_policy_schema());
    }
};
DecodePolicyRegistrar g_decode_policy_registrar{};
} // namespace
} // namespace trtf::config::schemas

// Force-link anchor.
namespace trtf::config::schemas {
volatile int kForceLink_decode_policy = 0;
} // namespace trtf::config::schemas

File: trtf_build/trtf_build/runtime_config/schemas/decode_policy.py
"""Schema for the ``decode_policy`` namespace.

Covers the build-time decoder-attention knobs previously read via
environment variables. The namespace is deliberately small — most runtime
decode behavior lives in other namespaces.
"""

from __future__ import annotations

from trtf_build.runtime_config import (
    ConfigField,
    Layer,
    Schema,
    register_schema,
)


_BUILD_AND_BUNDLE = frozenset({Layer.BUILD_TIME, Layer.BUNDLE_DEFAULT})


SCHEMA = Schema(
    namespace="decode_policy",
    fields=(
        # Replaces TRTF_FORCE_MANUAL_DECODER_ATTENTION (build-time env var).
        # Build-time only: chosen when the bundle is built and baked into
        # the engine graph — cannot change at run time. Session/platform
        # layers are not in the allowlist.
        ConfigField(
            name="force_manual_attention",
            type_tag="bool",
            default=False,
            allowed_layers=_BUILD_AND_BUNDLE,
        ),
    ),
)


register_schema(SCHEMA)
